protocol/messenger.go

```diff
@@ -917,6 +944,134 @@ func (m *Messenger) ChangeGroupChatName(ctx context.Context, chatID string, name
 	return &response, m.saveChat(chat)
 }
 
+func (m *Messenger) SendGroupChatInvitationRequest(ctx context.Context, chatID string, adminPK string,
+	message string) (*MessengerResponse, error) {
+	logger := m.logger.With(zap.String("site", "SendGroupChatInvitationRequest"))
+	logger.Info("Sending group chat invitation request", zap.String("chatID", chatID),
+		zap.String("adminPK", adminPK), zap.String("message", message))
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	var response MessengerResponse
+
+	invitationR := &GroupChatInvitation{
+		GroupChatInvitation: protobuf.GroupChatInvitation{
+			ChatId:              chatID,
+			IntroductionMessage: message,
+		},
+		From: types.EncodeHex(crypto.FromECDSAPub(&m.identity.PublicKey)),
+	}
+
+	encodedMessage, err := proto.Marshal(invitationR.GetProtobuf())
+	if err != nil {
+		return nil, err
+	}
+
+	spec := common.RawMessage{
+		LocalChatID:         adminPK,
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_GROUP_CHAT_INVITATION,
+		ResendAutomatically: true,
+	}
+
+	pkey, err := hex.DecodeString(adminPK[2:])
+	if err != nil {
+		return nil, err
+	}
+	// Safety check, make sure is well formed
+	adminpk, err := crypto.UnmarshalPubkey(pkey)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := m.processor.SendPrivate(ctx, adminpk, spec)
+	if err != nil {
+		return nil, err
+	}
+
+	spec.ID = types.EncodeHex(id)
+	spec.SendCount++
+	err = m.persistence.SaveRawMessage(&spec)
+	if err != nil {
+		return nil, err
+	}
+
+	response.Invitations = []*GroupChatInvitation{invitationR}
+
+	err = m.persistence.SaveInvitation(invitationR)
+	if err != nil {
+		return nil, err
+	}
+
+	return &response, nil
+}
+
+func (m *Messenger) GetGroupChatInvitations() ([]*GroupChatInvitation, error) {
+	return m.persistence.GetGroupChatInvitations()
+}
+
+func (m *Messenger) SendGroupChatInvitationRejection(ctx context.Context, invitationRequestID string) (*MessengerResponse, error) {
+	logger := m.logger.With(zap.String("site", "SendGroupChatInvitationRejection"))
+	logger.Info("Sending group chat invitation reject", zap.String("invitationRequestID", invitationRequestID))
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	invitationR, err := m.persistence.InvitationByID(invitationRequestID)
+	if err != nil {
+		return nil, err
+	}
+
+	invitationR.Rejection = true
+
```

Comment:
Looking at this code you pull the invitation request from the db, which contains the `From` / `source` of the original requester, and then set the `Rejection` flag to true.

So there is no need to overwrite the `From` field when handling an incoming rejection message?

Reply:
`From / source` is the PK of invitation author it's not changed